src/borg/archiver.py

```diff
@@ -696,6 +698,195 @@ def peek_and_store_hardlink_masters(item, matched):
 
     @with_repository()
     @with_archive
+    def do_export_tar(self, args, repository, manifest, key, archive):
+        """Export archive contents as a tarball"""
+        self.output_list = args.output_list
+
+        filter = None
+        if args.tar_filter == 'auto':
+            # Note that filter remains None if tarfile is '-'.
+            if args.tarfile.endswith('.tar.gz'):
+                filter = 'gzip'
+            elif args.tarfile.endswith('.tar.bz2'):
+                filter = 'bzip2'
+            elif args.tarfile.endswith('.tar.xz'):
+                filter = 'xz'
+            logger.debug('Automatically determined tar filter: %s', filter)
+        else:
+            filter = args.tar_filter
+
+        if args.tarfile == '-':
+            tarstream, tarstream_close = sys.stdout.buffer, False
+        else:
+            tarstream, tarstream_close = open(args.tarfile, 'wb'), True
+
+        if filter:
+            # When we put a filter between us and the final destination,
+            # the selected output (tarstream until now) becomes the output of the filter (=filterout).
+            # The decision whether to close that or not remains the same.
+            filterout = tarstream
+            filterout_close = tarstream_close
+            # There is no deadlock potential here (the subprocess docs warn about this), because
+            # communication with the process is a one-way road, i.e. the process can never block
+            # for us to do something while we block on the process for something different.
+            filterproc = subprocess.Popen(filter, stdin=subprocess.PIPE, stdout=filterout, shell=True)
+            # Always close the pipe, otherwise the filter process would not notice when we are done.
+            tarstream = filterproc.stdin
+            tarstream_close = True
+
+        # The | (pipe) symbol instructs tarfile to use a streaming mode of operation
+        # where it never seeks on the passed fileobj.
+        tar = tarfile.open(fileobj=tarstream, mode='w|')
+
+        self._export_tar(args, archive, tar)
+
+        # This does not close the fileobj (tarstream) we passed to it -- a side effect of the | mode.
+        tar.close()
+
+        if tarstream_close:
+            tarstream.close()
+
+        if filter:
+            logger.debug('Done creating tar, waiting for filter to die...')
+            rc = filterproc.wait()
+            logger.debug('filter exited with code %d', rc)
+            self.exit_code = max(self.exit_code, rc)
```

Comment:
gzip:

Exit  status  is normally 0; if an error occurs, exit status is 1. If a  warning occurs, exit status is 2.

bzip2:

       Return values: 0 for a normal exit, 1 for environmental problems  (file
       not found, invalid flags, I/O errors, &c), 2 to indicate a corrupt com‐
       pressed file, 3 for an  internal  consistency  error  (eg,  bug)  which
       caused bzip2 to panic.

xz:

       0      All is good.
       1      An error occurred.
       2      Something  worth  a  warning  occurred,  but  no  actual  errors
              occurred.